Tester/CodeCoverage/Generators/CloverXMLGenerator.php

```diff
@@ -0,0 +1,213 @@
+	);
+
+
+	public function __construct($file, $source = NULL)
+	{
+		if (!extension_loaded('dom')) {
+			throw new \LogicException("CloverXML generator requires DOM extension to be loaded.");
+		}
+		parent::__construct($file, $source);
+	}
+
+
+	protected function renderSelf()
+	{
+		$time = time();
+		$parser = new PhpParser;
+
+		$doc = new DOMDocument;
+		$doc->formatOutput = TRUE;
+
+		$elCoverage = $doc->appendChild($doc->createElement('coverage'));
+		$elCoverage->setAttribute('generated', $time);
+
+		// TODO: @name
+		$elProject = $elCoverage->appendChild($doc->createElement('project'));
+		$elProject->setAttribute('timestamp', $time);
+		$elProjectMetrics = $elProject->appendChild($doc->createElement('metrics'));
+
+		$projectMetrics = (object) array(
+			'packageCount' => 0,
+			'fileCount' => 0,
+			'linesOfCode' => 0,
+			'linesOfNonCommentedCode' => 0,
+			'classCount' => 0,
+			'methodCount' => 0,
+			'coveredMethodCount' => 0,
+			'statementCount' => 0,
+			'coveredStatementCount' => 0,
+			'elementCount' => 0,
+			'coveredElementCount' => 0,
+			'conditionalCount' => 0,
+			'coveredConditionalCount' => 0,
+		);
+
+		foreach ($this->getSourceIterator() as $file) {
+			$file = (string) $file;
+
+			$projectMetrics->fileCount++;
+
+			$coverageData = isset($this->data[$file]) ? $this->data[$file] : NULL;
+
+			// TODO: split to <package> by namespace?
+			$elFile = $elProject->appendChild($doc->createElement('file'));
+			$elFile->setAttribute('name', $file);
+			$elFileMetrics = $elFile->appendChild($doc->createElement('metrics'));
+
+			$code = $parser->parse(file_get_contents($file));
+
+			$fileMetrics = (object) array(
+				'linesOfCode' => $code->linesOfCode,
+				'linesOfNonCommentedCode' => $code->linesOfCode - $code->linesOfComments,
+				'classCount' => count($code->classes) + count($code->traits),
+				'methodCount' => 0,
+				'coveredMethodCount' => 0,
+				'statementCount' => 0,
+				'coveredStatementCount' => 0,
+				'elementCount' => 0,
+				'coveredElementCount' => 0,
+				'conditionalCount' => 0,
+				'coveredConditionalCount' => 0,
+			);
+
+			foreach (array_merge($code->classes, $code->traits) as $name => $info) { // TODO: interfaces?
+				$elClass = $elFile->appendChild($doc->createElement('class'));
+				if (($tmp = strrpos($name, '\\')) === FALSE) {
+					$elClass->setAttribute('name', $name);
+				} else {
+					$elClass->setAttribute('namespace', substr($name, 0, $tmp));
+					$elClass->setAttribute('name', substr($name, $tmp + 1));
+				}
+
+				$elClassMetrics = $elClass->appendChild($doc->createElement('metrics'));
+				$classMetrics = $this->calculateClassMetrics($info, $coverageData);
+				self::setMetricAttributes($elClassMetrics, $classMetrics);
+				self::appendMetrics($fileMetrics, $classMetrics);
+			}
+			self::setMetricAttributes($elFileMetrics, $fileMetrics);
+
+
+			foreach ((array) $coverageData as $line => $count) {
+				if ($count === -2) { // -2 = dead code
```

Comment:
What about to use constants, ie. AbstractGenerator::CODE_DEAD, CODE_TESTED and CODE_UNTESTED?